File: ruby/leap/year.rb

class Year

  def initialize(year)
    @year = year
  end

  def leap?
    vanilla_leap_year? || exceptional_century?
  end

  private

  def vanilla_leap_year?
    divisible_by?(4) && !century?
  end

  def century?
    divisible_by?(100)
  end

  def exceptional_century?
    divisible_by?(400)
  end

  def divisible_by?(n)
    @year % n == 0
  end

end

File: elixir/leap/year.exs
defmodule Year do
  @doc """
  Returns whether 'year' is a leap year.

  A leap year occurs:

  on every year that is evenly divisible by 4
    except every year that is evenly divisible by 100
      except every year that is evenly divisible by 400.
  """
  @spec leap_year?(non_neg_integer) :: boolean
  def leap_year?(year) do
    cond do
      _special_century?(year) -> true
      _century?(year)         -> false
      _divisible_by_4?(year)  -> true
      true                    -> false
    end
  end

  defp _special_century?(year), do: _divisible_by?(year, 400)
  defp _century?(year),         do: _divisible_by?(year, 100)
  defp _divisible_by_4?(year),  do: _divisible_by?(year, 4)
  defp _divisible_by?(year, n), do: rem(year, n) == 0

end
